File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/CurrentUserController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.dto.UserDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.UserSerializer
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/me")
class CurrentUserController(
    private val userSerializer: UserSerializer
) {
    @GetMapping
    fun get(@CurrentUser(required = false) currentUser: User?,
            @RequestParam(defaultValue = "{}") options: UserSerializer.Options): UserDTO {
        if (currentUser == null) {
            throw ApiException("Not logged in", HttpStatus.FORBIDDEN)
        }
        return userSerializer.serialize(currentUser, currentUser, options)
    }
}

File: backend-ts/src/controllers/current_user.controller.ts
import { Controller, DefaultValuePipe, Get, Query } from "@nestjs/common";
import { UserDTO } from 'shared/types';
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { UserSerializer, UserSerializerOptions } from "src/serializers/user.serializer";
import { ApiException } from "./exceptions";

@Controller('/api/v4/me')
export class CurrentUserController {
  constructor(
    private userSerializer: UserSerializer,
  ) {}

  @Get()
  async get(
    @Query('options', new DefaultValuePipe('{}')) optionsJson: string,
    @CurrentUser({ required: false }) currentUser: User | null,
  ): Promise<UserDTO> {
    if (!currentUser) throw new ApiException('Not logged in', 403)
    const options: UserSerializerOptions = JSON.parse(optionsJson)
    return this.userSerializer.serialize(currentUser, currentUser, options)
  }
}
